src/Icicle/Avalanche/Check.hs

```diff
@@ -88,7 +88,15 @@ checkStatement frag ctx stmt
                    $ checkExp frag (ctxExp ctx) x
                Let n x' <$> go stmts
 
-        ForeachInts n from to stmts
+        While t n from to stmts
+         -> do from' <- first ProgramErrorExp
+                      $ checkExp frag (ctxExp ctx) from
+               to'   <- first ProgramErrorExp
+                      $ checkExp frag (ctxExp ctx) to
+
+               While t n from' to' <$> go stmts
```

Comment:
it's a bigger change to typecheck what you're doing now though: you'd need to make the rest of the typechecking return the environment of bound scalars, and check that it has the same type as "from" etc.
whereas if you use an accumulator, you can just check that from and to types are the same, then insert "n" as an accumulator of that type and recurse